src/source/Signaling/LwsApiCalls.c

```diff
@@ -544,24 +544,56 @@ STATUS lwsCompleteSync(PLwsCallInfo pCallInfo)
 
     connectInfo.opaque_user_data = pCallInfo;
 
-    MUTEX_LOCK(pCallInfo->pSignalingClient->lwsServiceLock);
-    locked = TRUE;
-    CHK(NULL != lws_client_connect_via_info(&connectInfo), STATUS_SIGNALING_LWS_CLIENT_CONNECT_FAILED);
-    MUTEX_UNLOCK(pCallInfo->pSignalingClient->lwsServiceLock);
-    locked = FALSE;
+    // Attempt to iterate and acquire the locks
+    // NOTE: The https protocol should be called sequentially only
+    MUTEX_LOCK(pCallInfo->pSignalingClient->lwsSerializerLock);
+    serializerLocked = TRUE;
+
+    // Ensure we are not running another https protocol
+    CHK_ERR(pCallInfo->pSignalingClient->currentWsi[PROTOCOL_INDEX_HTTPS] == NULL, STATUS_INVALID_OPERATION,
+            "HTTPS requests should be processed sequentially.");
+    while (iterate && pCallInfo->pSignalingClient->currentWsi[PROTOCOL_INDEX_WSS] != NULL) {
```

Comment:
Would you mind to add some comments when the currentWsi for WSS is nullified? From my understanding, when we failed to acquire the service lock, we would wake up the thread that will eventually nullify the currentWsi for WSS. Is this right?

How do we make sure that the awaken thread will finish in timely manner?

Reply:
I added some wording in the comment. Basically, there is a lock contention. If there is an ongoing WSS operation indicated by a non-null wss pointer, and only then, we need to wake up the service thread handler which would release the lock that we could acquire. I realized that this will lead to a temporary busy loops while the contention of the lock is happening so I added an atomic to indicate to the awakening service loop to wait.

If the wss pointer is null then there is no need to acquire the lock.

Reply:
There could be two threads at one time driving the service loop - one that's exiting and the other that's either running reconnect or the https APIs. The changes I made will remove the potentially busy loop in case of thread starvation by adding THREAD_SLEEP. 

THREAD_SLEEP is a non-blocking sleep implemented by the OS scheduled. What this means, is that the thread is relinquishing it's quanta, allowing another competing thread to grab the lock.